nixos/modules/services/monitoring/prometheus/collectd-exporter.nix

```diff
@@ -0,0 +1,134 @@
+
+      port = mkOption {
+        type = types.int;
+        default = 9103;
+        description = ''
+          Port to listen on.
+          This is used for scraping as well as the to receive collectd data via the write_http plugin.
+        '';
+      };
+
+      listenAddress = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        example = "0.0.0.0";
+        description = ''
+          Address to listen on for web interface, telemetry and collectd JSON data.
+        '';
+      };
+
+      collectdBinary = mkOption {
+        description = ''
+          Configure collection via the collectd binary network protocol.
+        '';
+        type = (types.submodule {
+          options = {
+            enable = mkEnableOption "collectd binary protocol receiver";
+
+            authFile = mkOption {
+              default = null;
+              type = types.nullOr types.path;
+              description = "File mapping user names to pre-shared keys (passwords).";
+            };
+
+            port = mkOption {
+              type = types.int;
+              default = 25826;
+              description = ''Network address on which to accept collectd binary network packets.'';
+            };
+
+            listenAddress = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              example = "0.0.0.0";
+              description = ''
+                Address to listen on for binary network packets.
+              '';
+            };
+
+            securityLevel = mkOption {
+              type = types.enum ["None" "Sign" "Encrypt"];
+              default = "None";
+              description = ''
+                Minimum required security level for accepted packets.
+              '';
+            };
+          };
+        });
+      };
+
+      extraFlags = mkOption {
+        type = types.listOf types.str;
+        default = [];
+        description = ''
+          Extra commandline options when launching the collectd exporter.
+        '';
+      };
+
+      logFormat = mkOption {
+        type = types.str;
+        default = "logger:stderr";
+        example = "logger:syslog?appname=bob&local=7 or logger:stdout?json=true";
+        description = ''
+          Set the log target and format.
+        '';
+      };
+
+      logLevel = mkOption {
+        type = types.enum ["debug" "info" "warn" "error" "fatal"];
+        default = "info";
+        description = ''
+          Only log messages with the given severity or above.
+        '';
+      };
+
+      openFirewall = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Open port in firewall for incoming connections.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    networking.firewall.allowedTCPPorts = [ (optional cfg.openFirewall cfg.port) (optional (cfg.openFirewall && cfg.collectdBinary.enable) cfg.collectdBinary.port) ];
```

Comment:
Throws an arrow if `openFirewall` is false (default):

```
error: The option value `networking.firewall.allowedTCPPorts.[definition 1-entry 1]' in `/home/joerg/git/nixpkgs/nixos/modules/services/monitoring/prometheus/collectd-exporter.nix' is not a integer
```

Reply:
@Mic92 Do you know the correct way to enable the second port based on two conditions? I couldn't find any example.

Reply:
untested:

```nix
networking.firewall.allowedTCPPorts = (optional cfg.openFirewall cfg.port)  ++ 
  (optional (cfg.openFirewall && cfg.collectdBinary.enable) cfg.collectdBinary.port);
```

Reply:
I will give it a try. I now also figured out why I didn't catch the issue. I have the firewall disabled on my dev machine.

Reply:
@Mic92 Your suggestions works. But I was wondering if it would make more sense to have two firewall open rules one for the binary and one for the scrape/json. Maybe people only want to open one.

Reply:
I would keep it simple, you can always use `networking.firewall.allowedTCPPorts = [ ]` yourself, if you want it more granular.